src/js/node/fs.ts

```diff
@@ -535,341 +535,308 @@ var defaultReadStreamOptions = {
+  n =
+    pos !== undefined // if there is a pos, then we are reading from that specific position in the file
+      ? Math.min(end - pos + 1, n) // takes smaller of length of the rest of the file to read minus the cursor position, or the highwatermark
+      : Math.min(end - bytesRead + 1, n); // takes the smaller of the length of the rest of the file from the bytes that we have marked read, or the highwatermark
 
-      // If n is 0 or less, then we read all the file, push null to stream, ending it
-      if (n <= 0) {
-        this.push(null);
-        return;
-      }
+  $debug("n @ fs.ReadStream.#internalRead, after clamp", n);
 
-      // At this point, n is the lesser of the length of the rest of the file to read or the highwatermark
-      // Which means n is the maximum number of bytes to read
-
-      // Basically if we don't know the file size yet, then check it
-      // Then if n is bigger than fileSize, set n to be fileSize
-      // This is a fast path to avoid allocating more than the file size for a small file (is this respected by native stream though)
-      if (this.#fileSize === -1 && bytesRead === 0 && pos === undefined) {
-        var stat = fstatSync(fd);
-        this.#fileSize = stat.size;
-        if (this.#fileSize > 0 && n > this.#fileSize) {
-          n = this.#fileSize + 1;
-        }
-        $debug("fileSize", this.#fileSize);
-      }
+  // If n is 0 or less, then we read all the file, push null to stream, ending it
+  if (n <= 0) {
+    this.push(null);
+    return;
+  }
 
-      // At this point, we know the file size and how much we want to read of the file
-      this[kIoDone] = false;
-      var res = super._read(n);
-      $debug("res -- undefined? why?", res);
-      if ($isPromise(res)) {
-        var then = res?.then;
-        if (then && $isCallable(then)) {
-          res.then(
-            () => {
-              this[kIoDone] = true;
-              // Tell ._destroy() that it's safe to close the fd now.
-              if (this.destroyed) {
-                this.emit(kIoDone);
-              }
-            },
-            er => {
-              this[kIoDone] = true;
-              this.#errorOrDestroy(er);
-            },
-          );
-        }
-      } else {
-        this[kIoDone] = true;
-        if (this.destroyed) {
-          this.emit(kIoDone);
-          this.#errorOrDestroy(new Error("ERR_STREAM_PREMATURE_CLOSE"));
-        }
-      }
+  // At this point, n is the lesser of the length of the rest of the file to read or the highwatermark
+  // Which means n is the maximum number of bytes to read
+
+  // Basically if we don't know the file size yet, then check it
+  // Then if n is bigger than fileSize, set n to be fileSize
+  // This is a fast path to avoid allocating more than the file size for a small file (is this respected by native stream though)
+  if (this[mfileSize] === -1 && bytesRead === 0 && pos === undefined) {
+    var stat = fstatSync(fd);
```

Comment:
we should remove this code later, we already call stat in Bun.file(path).stream

not in this PR though